javalib/src/main/scala/java/math/BigInteger.scala

```diff
@@ -0,0 +1,1500 @@
+/*
+ *  Ported by Alistair Johnson from  https://github.com/gwtproject/gwt/blob/master/user/super/com/google/gwt/emul/java/math/BigInteger.java
+ */
+package java.math
+
+import java.io.Serializable
+import java.util.Random
+
+import scala.annotation.tailrec
+
+/**
+ * Companion object for {@code BigInteger}
+ */
+object BigInteger {
+
+  /**
+   * The {@code BigInteger} constant 1.
+   */
+  val ONE = new BigInteger(1, 1)
+
+  /**
+   * The {@code BigInteger} constant 10.
+   */
+  val TEN = new BigInteger(1, 10)
+
+  /**
+   * The {@code BigInteger} constant 0.
+   */
+  val ZERO = new BigInteger(0, 0)
+
+  /**
+   * The {@code BigInteger} constant 0 used for comparison.
+   */
+  val EQUALS = 0
+
+  /**
+   * The {@code BigInteger} constant 1 used for comparison.
+   */
+  val GREATER = 1
+
+  /**
+   * The {@code BigInteger} constant -1 used for comparison.
+   */
+  val LESS = -1
+
+  /**
+   * The {@code BigInteger} constant -1.
+   */
+  val MINUS_ONE = new BigInteger(-1, 1)
+
+  /**
+   * 2^32.
+   */
+  val POW32 = 4294967296d
+
+  /**
+   * All the {@code BigInteger} numbers in the range [0,10] are cached.
+   */
+  val SMALL_VALUES = Array(
+    ZERO, ONE, new BigInteger(1, 2), new BigInteger(1, 3),
+    new BigInteger(1, 4), new BigInteger(1, 5), new BigInteger(1, 6),
+    new BigInteger(1, 7), new BigInteger(1, 8), new BigInteger(1, 9), TEN)
+
+  val TWO_POWS = new Array[BigInteger](32)
+
+  for (i <- 0 until TWO_POWS.length) {
+    TWO_POWS(i) = BigInteger.valueOf(1L << i)
+  }
```

Comment:
You can use `Array.tabulate` to do this in a functional one-liner:
```scala
private val TWO_POWS = Array.tabulate[BigInteger](32)(i => BigInteger.valueOf(1 << i))
```
(note the `1` instead of `1L`, that will cause the `<<` to operate at the `Int` level, and then convert to `Long`, so it allocates one `RuntimeLong` fewer ^^)